src/Simulation/Core/IQirMachine.cs

```diff
@@ -0,0 +1,218 @@
+        public class Range : ArgumentValue
+        {
+            /// <summary>
+            /// The value of the argument.
+            /// </summary>
+            public QRange Value { get; }
+
+            /// <summary>
+            /// Creates a range argument value.
+            /// </summary>
+            /// <param name="value">The value of the argument.</param>
+            public Range(QRange value) => this.Value = value;
+        }
+
+        /// <summary>
+        /// A result argument value.
+        /// </summary>
+        public class Result : ArgumentValue
+        {
+            /// <summary>
+            /// The value of the argument.
+            /// </summary>
+            public Core.Result Value { get; }
+
+            /// <summary>
+            /// Creates a result argument value.
+            /// </summary>
+            /// <param name="value">The value of the argument.</param>
+            public Result(Core.Result value) => this.Value = value;
+        }
+
+        /// <summary>
+        /// A string argument value.
+        /// </summary>
+        public class String : ArgumentValue
+        {
+            /// <summary>
+            /// The value of the argument.
+            /// </summary>
+            public string Value { get; }
+
+            /// <summary>
+            /// Creates a string argument value.
+            /// </summary>
+            /// <param name="value">The value of the argument.</param>
+            public String(string value) => this.Value = value;
+        }
+
+        /// <summary>
+        /// An array argument value where all values are of the same type.
+        /// </summary>
+        public class Array : ArgumentValue
+        {
+            /// <summary>
+            /// The values of the argument.
+            /// </summary>
+            public IQArray<ArgumentValue> Values { get; }
+
+            private Array(IQArray<ArgumentValue> values) => this.Values = values;
+
+            /// <summary>
+            /// Tries to create an array argument value.
+            /// </summary>
+            /// <param name="values">The values of the argument.</param>
+            /// <returns>The array or <c>null</c> if not all values are of the same type.</returns>
+            public static Array? TryCreate(IQArray<ArgumentValue> values) => 
+                IsHomogeneous(values) ? new Array(values) : null;
+
+            private static bool IsHomogeneous(IQArray<ArgumentValue> values) => values
+                .Zip(values.Skip(1), ValueTuple.Create)
+                .All(pair => pair switch 
+                {
+                    (Bool _, Bool _) => true,
+                    (Int _, Int _) => true,
+                    (Double _, Double _) => true,
+                    (Pauli _, Pauli _) => true,
+                    (Range _, Range _) => true,
+                    (Result _, Result _) => true,
+                    (String _, String _) => true,
+                    (Array _, Array _) => true,
```

Comment:
I think this won't prevent arrays of arrays from having different item types - I'll replace this with a better check.